test/integration/api/public-links.js

```diff
@@ -0,0 +1,231 @@
+          .then(({ body }) => asBob.delete('/v1/projects/1/forms/simple/public-links/' + body.id).expect(200)))));
+
+    it('should only delete the token if it is part of the form', testService((service) =>
+      service.login('alice', (asAlice) =>
+        asAlice.post('/v1/projects/1/forms/withrepeat/public-links')
+          .send({ displayName: 'condemned' })
+          .expect(200)
+          .then(({ body }) => asAlice.delete(`/v1/projects/1/forms/simple/public-links/${body.id}`)
+            .expect(404)))));
+  });
+});
+
+
+// Test the actual use of public links.
+describe('api: /key/:key', () => {
+  it('should return 401 if an invalid key is provided', testService((service) =>
+    service.get('/v1/key/aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa/users/current')
+      .expect(401)));
+
+  it('should return 401 if two credentials are presented', testService((service) =>
+    service.login('alice', (asAlice) => asAlice.post('/v1/projects/1/forms/simple/public-links')
+      .send({ displayName: 'linktest' })
+      .then(({ body }) => asAlice.get(`/v1/key/${body.token}/users/current`)
+        .expect(401)))));
+
+  it('should allow cookie+public-link', testService((service) =>
+    service.post('/v1/sessions')
+      .send({ email: 'alice@opendatakit.org', password: 'alice' })
+      .expect(200)
+      .then(({ body }) => body.token)
+      .then((aliceToken) => service.login('alice', (asAlice) =>
+        asAlice.post('/v1/projects/1/forms/simple/public-links')
+          .send({ displayName: 'linktest' })
+          .then(({ body }) => body.token)
+          .then((linkToken) => service.post(`/v1/key/${linkToken}/projects/1/forms/simple/submissions`)
+            .send(testData.instances.simple.one)
+            .set('Content-Type', 'application/xml')
+            .set('Cookie', `__Host-session=${aliceToken}`)
+            .expect(200))))));
+
+  it('should passthrough to the appropriate route with successful auth', testService((service) =>
+    service.login('alice', (asAlice) =>
+      asAlice.post('/v1/projects/1/forms/simple/public-links')
+        .send({ displayName: 'linktest' })
+        .then(({ body }) => body)
+        .then((link) => service.post(`/v1/key/${link.token}/projects/1/forms/simple/submissions`)
+          .send(testData.instances.simple.one)
+          .set('Content-Type', 'application/xml')
+          .expect(200)))));
+
+  it('should not allow creating submissions on other forms', testService((service) =>
+    service.login('alice', (asAlice) =>
+      asAlice.post('/v1/projects/1/forms/simple/public-links')
+        .send({ displayName: 'linktest' })
+        .then(({ body }) => body)
+        .then((link) => service.post(`/v1/key/${link.token}/projects/1/forms/withrepeat/submissions`)
+          .send(testData.instances.withrepeat.one)
+          .set('Content-Type', 'application/xml')
+          .expect(403)))));
```

Comment:
This looks right. However, it looks like if a retired public link is used, there'll be a 401. That's going to lead to confusing end-user behavior, I think. Let's say a respondent starts filling out the form and then the public link is retired. The respondent will be asked to authenticate on submission which is entirely unexpected. The best option I can think of is to give the 401 so that Enketo redirects but then have Central identify that a revoked token is being used and provide a message accordingly. We can probably leave as-is for now but I think we'll want an answer pretty soon unless I've misunderstood what happens.

I'm having trouble figuring out what happens with `formList` requests. I tried to write a test for that but I always get an empty response so I must be doing it wrong. I would expect the single form that the link is built on to be in the list. I believe that in the case of an entirely wrong token, the formList request will 404 which seems ok. However, I suspect it will 401 in the case of a retired public link which would again be really surprising user behavior.

Thinking through all this has revealed some unpleasantness around error cases. I think some of the Enketo strings could be improved and will work with Martijn on that.

Reply:
Currently Enketo gets a 403 when requesting or submitting to a link with an expired or never-existed token. The error messages are not helpful. Enketo issue at https://github.com/enketo/enketo-express/issues/463

Reply:
Looks like it's probably a 401 for submission: https://github.com/getodk/central-backend/pull/278/commits/7b13eb7ef10cb964f17d16e0f2045b60c7bf3ea0